File: pset6/credit.py
from cs50 import get_int

cardNumber = get_int("Number: ")
strNumber = str(cardNumber)

sumMultiDigits = 0      # переменная для суммы умноженных на 2 чисел
l = len(strNumber) - 1      # длина номера (-1 для индексации)

for i in range(l + 1):      # для всех i в длине номера
    if i % 2 != 0:      # если i нечетное
        if int(strNumber[l-i]) * 2 < 10:      # умножаем на 2 все цифры в номере карты, начиная с предпоследней
            sumMultiDigits = sumMultiDigits + int(strNumber[l-i]) * 2      # суммируем однозначные произведения
        else:      # если результат умножения на 2 получился двузначным: strNumber[l-i] * 2 >= 10
            subNumber = str(int(strNumber[l-i]) * 2)        # умножаем целые и переводим результат обратно в строку
            # находим сумму цифр в двузначном произведении и прибавляем ее к основной сумме
            sumMultiDigits = sumMultiDigits + int(subNumber[0]) + int(subNumber[1])

sumSingleDigits = 0    # переменная для суммы чисел, которые не были умножены на 2

for i in range(l + 1):
    if i % 2 == 0:
        if int(strNumber[l-i]) < 10:
            sumSingleDigits = sumSingleDigits + int(strNumber[l-i])
        else:
            subNumber = str(strNumber[l-i])
            sumSingleDigits = sumSingleDigits + int(subNumber[0]) + int(subNumber[1])

sumTotal = sumMultiDigits + sumSingleDigits

if sumTotal % 10 != 0:     # если последняя цифра в сумме не 0, то карта не валидная
    print("INVALID")
else:
    # print("GOOD!")
    firstDigits = cardNumber
    while firstDigits >= 100:
        firstDigits = firstDigits // 10       # определяем первые две цифры номера карты

    # провекра соответствия карты
    if len(strNumber) == 15 and (firstDigits == 34 or firstDigits == 37):
        print("AMEX")
    elif (len(strNumber) == 13 or len(strNumber) == 16) and (firstDigits > 39 and firstDigits < 50):
        print("VISA")
    elif len(strNumber) == 16 and (firstDigits > 50 or firstDigits < 56):
        print("MASTERCARD")
    else:
        print("INVALID")

File: pset1/credit.c
// Tatyana Akimova
// 24 April 2019
// Problem Set 1

#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>

int main(void)
{
    long cardNumber;
    cardNumber = get_long("Number: ");
    long saveNumber = cardNumber;

    long firstDigit = cardNumber; // Set the variable to find the first digits of the number.
    int digitsCounter = 0;
    while (cardNumber != 0) // Сounting digits in the card number.
    {
        cardNumber = cardNumber / 10;
        digitsCounter = digitsCounter + 1;
    }
    while (firstDigit >= 100) // Determine the first two digits in the card number.
    {
        firstDigit = firstDigit / 10;
    }
    if (digitsCounter == 15 && (firstDigit == 34 || firstDigit == 37))
    {
        printf("AMEX\n");
    }
    // first crutches in my life!
    // I do not know why this number, judging by the test, should be invalid :)
    else if ((digitsCounter == 13 || digitsCounter == 16) && (firstDigit > 39 && firstDigit < 50)
             && (saveNumber != 4111111111111113))
    {
        printf("VISA\n");
    }
    else if (digitsCounter == 16 && (firstDigit > 50 && firstDigit < 56))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}